### PR TITLE
SLVUU-49: Improve Cypress Selectors

### PR DESCRIPTION
**Pull request by pling-scottlogic**
### Description
There is currently one Cypress test for the layout persistence epic, which uses brittle selectors such as CSS class names. The changes here improve the existing selectors to be more accessible and represent user interaction more accurately. We also establish a more robust testing pattern by way of the Page Object Model (POM), in which more accessible selectors are included to allow for future end-to-end testing of the layout management user journey.

### Change List
- Introduce Page Object Model (POM), abstracting accessible selectors for:
  - "ShellWithNewTheme" showcase example
  - Save Layout Dialog
- Add accessible properties to the following components to improve selectors:
  - `Stack` (`aria-label`)
  - `LayoutList` (`role`)
  - `Tabstrip` (`role`)
- Refactor existing Cypress test to use POM
- Rename exported `LayoutList` component to match filename
- Add "task" to Cypress config to allow console logging from tests

### Testing
Temporarily inserted all new POM methods into screenshot test to ensure selectors are working as expected.

### CODE REVIEW

File: vuu-ui/cypress/e2e/layout-management/screenshot.cy.js [hunk 1]
```diff
@@ -1,29 +1,25 @@
...
+const page = new ShellWithNewTheme();
+const dialog = new SaveLayoutDialog();
 
 context("Screenshot", () => {
   beforeEach(() => {
-    cy.visit(SHELL_WITH_NEW_THEME_URL);
+    page.visit();
   });
 
   // TODO (#VUU24): Improve test alignment with the user flow
```
Comment (pling-scottlogic) [inline vuu-ui/cypress/e2e/layout-management/screenshot.cy.js:13]: Can we remove this comment now?

File: vuu-ui/cypress/pages/ShellWithNewTheme.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    return cy.findByRole("tab", { name: "MY LAYOUTS" });
+  }
+
+  getSavedLayoutButton(layoutName, creator, date) {
+    const day = ("00" + date.getDate()).slice(-2);
+    const formattedDate = `${day}.${date.getMonth() + 1}.${date.getFullYear()}`;
+    const elementName = `${layoutName} ${creator}, ${formattedDate}`;
+
+    return cy.findByRole("button", { name: elementName });
+  }
```
Comment (pling-scottlogic) [inline vuu-ui/cypress/pages/ShellWithNewTheme.js:30]: I feel like there must be a better way to do this. Suggestions welcome.

Comment (vferraro-scottlogic) [reply vuu-ui/cypress/pages/ShellWithNewTheme.js:30]: I think here we can use formatDate from vuu-utils like such: `const formattedDate = formatDate(date, "dd.mm.yyyy")`

File: vuu-ui/cypress/e2e/layout-management/screenshot.cy.js (hunk 1 above)
Comment (cfisher-scottlogic) [reply vuu-ui/cypress/e2e/layout-management/screenshot.cy.js:13]: Yeah we've got plenty of tickets to cover this off now.

File: vuu-ui/cypress/pages/ShellWithNewTheme.js (hunk 2 above)
Comment (cfisher-scottlogic) [reply vuu-ui/cypress/pages/ShellWithNewTheme.js:30]: In terms of a user point of view, I would imagine the ideal flow would be something more along the lines of what you have in `getContextMenuButton`. So `get all layouts in my layouts -> find group x, find the one with y name or z date`.

File: vuu-ui/cypress/pages/ShellWithNewTheme.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import { SHELL_WITH_NEW_THEME_URL } from "../support/e2e/constants";
+
+export class ShellWithNewTheme {
+  visit() {
+    cy.visit(SHELL_WITH_NEW_THEME_URL);
+  }
+
+  getContextMenuButton() {
+    return cy
+      .findByRole("tablist", { name: "tab headers" })
```
Comment (cfisher-scottlogic) [inline vuu-ui/cypress/pages/ShellWithNewTheme.js:10]: I see you've added this `aria-label`, but what is "tab headers"? I'm struggling to relate this to the page.

File: vuu-ui/cypress/pages/ShellWithNewTheme.js (hunk 2 above)
Comment (pling-scottlogic) [reply vuu-ui/cypress/pages/ShellWithNewTheme.js:30]: @vferraro-scottlogic Thanks, that's exactly what I was looking for!

@cfisher-scottlogic I think that would involve a bit of refactor in the `LayoutList` component, but it should be doable. I'll look into it.

File: vuu-ui/cypress/pages/ShellWithNewTheme.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+      .findAllByRole("tab")
+      .first()
+      .findByRole("button");
```
[the remark this answers is not in the document]
Comment (pling-scottlogic) [reply vuu-ui/cypress/pages/ShellWithNewTheme.js]: 42c04a48

File: vuu-ui/cypress/pages/ShellWithNewTheme.js (hunk 3 above)
Comment (pling-scottlogic) [reply vuu-ui/cypress/pages/ShellWithNewTheme.js]: I went with just `layouts` in the end (42c04a48). I figured that label - in combination with the existing role ("tab list") - would be enough to specify it uniquely.

Comment (pling-scottlogic) [reply vuu-ui/cypress/pages/ShellWithNewTheme.js]: As discussed, went with `data tabs` (8acc2513)

**Review by cfisher-scottlogic — APPROVED** [review]

**Review by vferraro-scottlogic — CHANGES_REQUESTED** [review]
Only a couple of questions and some recommendations around leveraging typescript in cypress code

File: vuu-ui/cypress/pages/SaveLayoutDialog.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
+export class SaveLayoutDialog {
```
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply vuu-ui/cypress/pages/SaveLayoutDialog.ts:1]: I think for most of these the return type will be `Cypress.Chainable<JQuery<HTMLElement>>`

[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply vuu-ui/cypress/pages/SaveLayoutDialog.ts:1]: this is the difference at compile-time
(i.e. in screenshot.cy.ts)
before
after

File: vuu-ui/cypress/pages/SaveLayoutDialog.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  getScreenshot() {
+    return cy.findByAltText("screenshot of current layout");
+  }
+
+  getGroupField() {
+    return cy.findAllByRole("combobox", { name: "Group" });
```
[the remark this answers is not in the document]
Comment (pling-scottlogic) [reply vuu-ui/cypress/pages/SaveLayoutDialog.ts]: Good spot. I've changed it to `findByRole`.

File: vuu-ui/cypress/pages/SaveLayoutDialog.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (pling-scottlogic) [reply vuu-ui/cypress/pages/SaveLayoutDialog.ts:1]: afb79925

File: vuu-ui/packages/vuu-shell/src/layout-management/LayoutList.tsx [hunk 7]
```diff
@@ -43,32 +48,39 @@ export const LayoutsList = (props: HTMLAttributes<HTMLDivElement>) => {
...
-                    </div>
-                    <div className={`${classBase}-layoutDetails`}>
-                      <div>{`${metadata?.user}, ${metadata?.created}`}</div>
-                    </div>
-                  </div>
-                </div>
-              ))}
-            </>
+              {layoutMetadata.map((metadata) =>
+                LayoutTile(metadata, handleLoadLayout)
```
[the remark this answers is not in the document]
Comment (pling-scottlogic) [reply vuu-ui/packages/vuu-shell/src/layout-management/LayoutList.tsx]: No reason. I've converted it into a proper component and extracted into a separate file.

**Review by vferraro-scottlogic — APPROVED** [review]
Looks good :)

**Comment by pling-scottlogic** [comment]
[PR raised into Finos repo](https://github.com/finos/vuu/pull/1079)

**Comment by pling-scottlogic** [comment]
> [PR raised into Finos repo](https://github.com/finos/vuu/pull/1079)

Finos PR merged into main.

**Pull request closed without merging**